parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java

```
@@ -0,0 +1,330 @@
+    private Map<Integer, BloomFilter> bloomCache = null;
+    private Map<Integer, Type> types = null;
+    private boolean hasBloom = false;
+
+    private boolean eval(MessageType fileSchema, BlockMetaData rowGroup, BloomFilterReader bloomFilterReader) {
+      this.bloomReader = bloomFilterReader;
+      this.fieldsWithNoBloom = Sets.newHashSet();
+      this.cols = Maps.newHashMap();
+      this.columnMetaMap = Maps.newHashMap();
+      this.types = Maps.newHashMap();
+
+      for (ColumnDescriptor desc : fileSchema.getColumns()) {
+        PrimitiveType colType = fileSchema.getType(desc.getPath()).asPrimitiveType();
+        if (colType.getId() != null) {
+          int id = colType.getId().intValue();
+          cols.put(id, desc);
+        }
+      }
+
+      for (ColumnChunkMetaData meta : rowGroup.getColumns()) {
+        PrimitiveType colType = fileSchema.getType(meta.getPath().toArray()).asPrimitiveType();
+        if (colType.getId() != null) {
+          int id = colType.getId().intValue();
+          Type icebergType = schema.findType(id);
+          boolean noBloom = ParquetUtil.hasNoBloomFilterPages(meta);
+          if (!noBloom) {
+            hasBloom = true;
+          } else {
+            fieldsWithNoBloom.add(id);
+          }
+          columnMetaMap.put(id, meta);
+          types.put(id, icebergType);
+        }
+      }
+      if (!hasBloom) {
+        return ROWS_MIGHT_MATCH;
+      }
+      return ExpressionVisitors.visitEvaluator(expr, this);
+    }
+
+    @Override
+    public Boolean alwaysTrue() {
+      return ROWS_MIGHT_MATCH; // all rows match
+    }
+
+    @Override
+    public Boolean alwaysFalse() {
+      return ROWS_CANNOT_MATCH; // all rows fail
+    }
+
+    @Override
+    public Boolean not(Boolean result) {
+      // not() should be rewritten by RewriteNot
+      // bloom filter is based on hash and cannot eliminate based on not
+      return ROWS_MIGHT_MATCH;
```

Since this calls `rewriteNot` on the filter, you may want to throw an `UnsupportedOperationException`. This is okay as well, but I don't think this will ever be hit.

---

Fixed. Thanks!